x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/support/DLSRoleQueryValidator.java

```diff
@@ -0,0 +1,177 @@
+        if (read != expected) {
+            throw new XContentParseException(parser.getTokenLocation(),
+                "expected [" + expected + "] but found [" + read + "] instead");
+        }
+    }
+
+    /**
+     * Evaluates the query if it is a template and then validates the query by parsing
+     * and building the {@link QueryBuilder}. It also checks if the query type is
+     * supported in DLS role query.
+     *
+     * @param query            {@link BytesReference} query field from the role
+     * @param scriptService    {@link ScriptService} used for evaluation of a template query
+     * @param xContentRegistry {@link NamedXContentRegistry} for finding named queries
+     * @param user             {@link User} used when evaluation a template query
+     * @return {@link QueryBuilder} if the query is valid and allowed, in case {@link RoleDescriptor.IndicesPrivileges}
+     * * does not have a query field then it returns {@code null}.
+     */
+    @Nullable
+    public static QueryBuilder validateAndVerifyRoleQuery(BytesReference query, ScriptService scriptService,
+                                                          NamedXContentRegistry xContentRegistry, User user) {
+        QueryBuilder queryBuilder = null;
+        if (query != null) {
+            try {
+                String templateResult = SecurityQueryTemplateEvaluator.evaluateTemplate(query.utf8ToString(), scriptService,
+                    user);
+                queryBuilder = validateAndVerifyRoleQuery(templateResult, xContentRegistry);
+            } catch (ElasticsearchParseException | ParsingException | XContentParseException | IOException e) {
+                throw new ElasticsearchParseException("failed to parse field 'query' from the role descriptor", e);
+            }
+        }
+        return queryBuilder;
+    }
+
+    private static QueryBuilder validateAndVerifyRoleQuery(String query, NamedXContentRegistry xContentRegistry) {
+        QueryBuilder queryBuilder = null;
+        if (query != null) {
+            try {
+                try (XContentParser parser = XContentFactory.xContent(query).createParser(xContentRegistry,
+                    LoggingDeprecationHandler.INSTANCE, query)) {
+                    queryBuilder = AbstractQueryBuilder.parseInnerQueryBuilder(parser);
+                    verifyRoleQuery(queryBuilder);
+                }
+            } catch (ElasticsearchParseException | ParsingException | XContentParseException | IOException e) {
+                throw new ElasticsearchParseException("failed to parse field 'query' from the role descriptor", e);
+            }
+        }
+        return queryBuilder;
+    }
+
+    /**
+     * Checks whether the role query contains queries we know can't be used as DLS role query.
+     *
+     * @param queryBuilder {@link QueryBuilder} for given query
+     */
+    public static void verifyRoleQuery(QueryBuilder queryBuilder) {
```

Comment:
Sure thing. I mentioned this as it wasn't obvious in the beginning of the review that this was not introduced in this PR, but left it as a comment so we can see if this needs to be taken up in a separate one. If you agree that we should tackle it, could you open an issue to keep track so that we don't forget ?